### DISCUSSION TITLE
Is there an analogue of the withObservationTracking for use outside View?

### DISCUSSION THREAD

**Discussion by riley-usagi** [pointfreeco/swift-perception#22] [Q&A]
Hello.

Tell me, please, is it possible to track changes in the parameters outside the View hierarchy?

**Comment by mbrandonw** [comment]
Hi @riley-usagi, yes, we provide a `withPerceptionTracking` method that works exactly as Swift's native `withObservationTracking`.

**Reply by PatrickDanino** [reply]
Keep in mind that you need to be careful using `withPerceptionTracking` from within a class. After some time spent debugging, I filed this Radar with Apple: https://openradar.appspot.com/FB13556571

TL,DR: `withPerceptionTracking` seems to break SwiftUI views if you reference a property from within the class that is not itself a `Perceptible/Observable` object. Here's a very simple example that breaks any SwiftUI view tracking `count`:

```
@Perceptible class Test {
  var count: Int = 0
  init() {
     withPerceptionTracking { _ = self.count }
     onChange: { [weak self] in print("Count: \(self?.count ?? 0)") }
  }
}
```

A view that is displaying an instance of `Test.count` will not work. If you comment out the code inside `init`, it will work. That wasn't intuitive at all.

I'll note that I was able to repro this on iOS 17 using `Observable` exclusively, so this is not due to the `Perception` framework.

The workaround, which worked for me, was to rely on `didSet` instead. You can also use `withPerceptionTracking` to track changes for properties that support `Perceptible/Observable`, similar to the example Apple has in the [SDK docs](https://developer.apple.com/documentation/observation/withobservationtracking(_:onchange:)).

**Reply by mbrandonw** [reply]
Hi @PatrickDanino, this doesn't have to do with `withObservationTracking` or `withPerceptionTracking`. It can happen without:

```swift
@Perceptible class Test {
  var count: Int = 0
  init() {
    _ = self.count
  }
}
```

This is just a gotcha with using `@Observable` in general. If you access an observable field in the `init`, and if the class is created in a parent view's body, then that parent view will observe the field. We filed the issue with Apple and they consider it to be working as expected, although it is surprising.
